Simple payments: Show tax breakup in summary

selanthiraiyan opened the pull request:
Closes: #5809

### Description

#### What?
1. Get the tax breakup details (`tax_lines`) from the server and display it in the simple payments summary screen. (when `taxLinesInSimplePayments` feature flag is turned on)
2. Show/hide the "Charge Taxes" toggle based upon the tax configuration in `wp-admin`. i.e. If `wp-admin` is configured to charge no taxes, then we should hide the "Charge Taxes" toggle.

#### Why?
The tax percentage displayed in simple payment's summary screen is calculated locally on device and is inaccurate. To fix the inaccuracy and to provide more detailed tax information, we decided to get the detailed tax breakup details from the server and display it. 

#### How?
1. We receive the tax breakup details in the POST Order response under `tax_lines`.  We have stored the `tax_lines` information along with the `Order` (work completed [in past PRs](https://github.com/woocommerce/woocommerce-ios/issues/5809#issuecomment-1014552642)). In this PR, we use the `tax_lines` information of `Order` to display the tax breakup details in the summary screen. 
2. We will receive empty `tax_lines` if the merchant configured the taxes in `wp-admin` to not charge any taxes. In that case, we will hide the "Charge Taxes" toggle in the summary screen.

More details at p91TBi-7cZ-p2

### Changes
#### SimplePaymentsSummaryViewModel
- Introduce `showTaxBreakup` property to decide whether to show the tax breakup details or not. (`taxLinesInSimplePayments` feature flag decides)
- Introduce `showChargeTaxesToggle` property to decide whether to show/hide the “Charge Taxes” toggle. (Hides when the `TaxLine`s are empty)
- Introduce `TaxLine` struct to store the `tax_lines` information.
- Generate  `TaxLine`s from `Order`'s `taxes` property.
- Not appending any string like "Tax" to the `label` received from the server. Because the merchant would already have the "tax" appended to the `label` in `wp-admin`. (In their preferred language) For example, "State Tax" received from the server might end up into "State Tax Tax" if we try to manipulate the label on the device. It will get weird for labels configured in other languages.

#### SimplePaymentsSummaryViewModelTests
- Add tests to validate tax line related changes.

#### SimplePaymentsSummaryView
- Show the tax breakup information if the feature flag is turned on.
- Hide the “Charge Taxes” toggle if there are no `tax_lines`.

### Testing instructions

#### Scenario 1
**Taxes configured in `wp-admin`**
1. Configure taxes in wp-admin.
1. Navigate to `Settings` and make sure that `Order creation` is turned on under `Experimental Features`
1. Navigate to `Orders` tab.
1. Tap on `+` button on top right and select `Simple payment` from the action sheet.
1. Enter an amount and tap `Next`.
1. Turn on "Charge Taxes" toggle and observe the tax breakup information being displayed.
1. Validate that the tax breakup details match the details from `wp-admin`.
1. Turn off "Charge Taxes" toggle and observe that the tax breakup information is hidden.

#### Scenario 2
**No taxes configured in `wp-admin`**
1. Configure `wp-admin` to charge no taxes. (Delete all taxes.)
1. Navigate to `Settings` and make sure that `Order creation` is turned on under `Experimental Features`
1. Navigate to `Orders` tab.
1. Tap on `+` button on top right and select `Simple payment` from the action sheet.
1. Enter an amount and tap `Next`.
1. Observe that no tax information is presented as `wp-admin` is configured to charge no taxes.
1. Also, observe that the "Charge Taxes" toggle is hidden.

### Screenshots

#### Tax configuration in `wp-admin`

<img width="1586" alt="Screenshot 2022-01-22 at 7 54 22 AM" src="https://user-images.githubusercontent.com/524475/150621294-60d7278e-3a6c-4a83-a48e-37dad8aad5df.png">

| Tax breakup when `taxLinesInSimplePayments` feature flag is ON | Dark | Light |
| --- | --- | --- |
| When `tax_lines` is not empty and </br> when "Charge Taxes" toggle is turned on </br></br> We show the tax breakup details| <img src="https://user-images.githubusercontent.com/524475/150616615-a2348ced-b90a-49a2-a463-1e632550649b.png" width="350"> | <img src="https://user-images.githubusercontent.com/524475/150616622-c32ff148-7a52-4182-b3a4-6651ed7f3fcc.png"  width="350"> |
| When `tax_lines` is not empty and </br> when "Charge Taxes" toggle is turned off </br></br> We hide the tax breakup details | <img src="https://user-images.githubusercontent.com/524475/150616974-808c7635-7c30-474c-803f-aa23ce09d4cf.png" width="350"> | <img src="https://user-images.githubusercontent.com/524475/150616983-f3a1d22f-6b96-4f76-a455-c699d8f01635.png"  width="350"> |
| When `tax_lines` is empty </br></br> We hide the "Charge Taxes" toggle | <img src="https://user-images.githubusercontent.com/524475/150616798-9b797563-2c68-4ceb-86c8-a27ea04674e5.png" width="350"> | <img src="https://user-images.githubusercontent.com/524475/150616793-6e51fd0a-e50f-4884-a14a-f37a524a3d9c.png"  width="350"> |

#### Large fonts 
| Dark | Light |
| --- | --- |
| <img src="https://user-images.githubusercontent.com/524475/150616383-23f008ac-d63f-4844-8f8b-143a6118022d.png" width="350"> | <img src="https://user-images.githubusercontent.com/524475/150616396-d417cf29-0657-4c72-b1d2-a67ccbed10ca.png"  width="350"> |

#### Landscape
| Dark | Light |
| --- | --- |
| <img src="https://user-images.githubusercontent.com/524475/150626985-096fe9f2-6c42-4047-99dc-33e10493696e.png" width="350"> | <img src="https://user-images.githubusercontent.com/524475/150626983-059c7e66-4257-4923-9e3a-6fc0fd9c539f.png"  width="350"> |

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Commit:
- Introduce `TaxLine` struct in `SimplePaymentsSummaryViewModel` to hold tax breakup information.

Commit:
- Add comment for `TaxLine`'s `id` property.

Commit:
SimplePaymentsSummaryViewModel `Localization`
- Add localized string "Tax" to generate `TaxLine`s.

Commit:
SimplePaymentsSummaryViewModel
- Generate `TaxLine`s to represent `taxes` inside `View`

Commit:
- Introduce `taxLines` property to store tax information and update SwiftUI previews and tests with the new parameter.

Commit:
Merge branch 'issue/5809-tax-lines-in-yosemite' into issue/5809-tax-summary

Commit:
SimplePaymentsSummaryViewModel
- Add `shouldShowTaxBreakup` property to check `taxLinesInSimplePayments` feature flag.

Commit:
- Add support for mocking `taxLinesInSimplePayments` flag in `MockFeatureFlagService`.

Commit:
- Fix whitespace warnings in `SimplePaymentsSummary`.

Commit:
MockFeatureFlagService
- Fix name of `isTaxLinesInSimplePaymentsOn` property.

Commit:
SimplePaymentsSummaryViewModelTests
- Add tests to validate `shouldShowTaxBreakup` behaviour.

Commit:
SimplePaymentsSummaryViewModelTests
- Stop creating `taxLines` manually as the current tests don't rely on `SimplePaymentsSummaryViewModel` having `taxLines`.

Commit:
SimplePaymentsSummaryViewModelTests
- Add tests to validate the process of the generating `taxLines` from `Order`'s `taxes`.

Commit:
Stop appending "Tax" to the `taxLine`'s `label`.
- The `label` is configured at wp-admin and the merchant might have "Tax" already appended to it. (In their preferred language)

Commit:
`shouldShowChargeTaxesToggle` in `SimplePaymentsSummaryViewModel`
- Add `shouldShowChargeTaxesToggle` to decide whether to show the "Charge Taxes" toggle.
- Add tests to validate `shouldShowChargeTaxesToggle` behaviour.

Commit:
- Rename variables to maintain naming convention inside `SimplePaymentsSummaryViewModel`

Commit:
SimplePaymentsSummary view
- Show/hide "Charge Taxes" toggle based on ViewModel's `showChargeTaxesToggle` value.

Commit:
SimplePaymentsSummary view
- Show tax breakup if ViewModel's `showTaxBreakup` is true.

Commit:
Merge branch 'trunk' into issue/5809-tax-summary

Commit:
Merge branch 'trunk' into issue/5809-tax-summary

Commit:
- Update release notes.

Commit:
- Stop having separate function to calculate taxLine in `SimplePaymentsSummary_Preview`

Commit:
TaxLine title
- Simplify `TaxLine`'s `title` generation to improve readability.
- Update tests.

Commit:
- Add comment to explain `test_taxLines_order_matches_taxes_order` test method.

itsmeichigo commented on RELEASE-NOTES.txt:
```diff
@@ -6,6 +6,7 @@
 - [***] We introduced a new tab called "Menu", a tab in the main navigation where you can browser different sub-sections of the app: Switch Store, Settings, WooCommerce Admin, View Store and Reviews. [https://github.com/woocommerce/woocommerce-ios/pull/5926]
 - [***] Store admins can now access sites with plugins that have Jetpack Connection Package (e.g. WooCommerce Payments, Jetpack Backup) in the app. These sites do not require Jetpack-the-plugin to connect anymore. Store admins can still install Jetpack-the-plugin from the app through settings or a Jetpack banner. [https://github.com/woocommerce/woocommerce-ios/pull/5924]
 - [*] Add/Edit Product screen: Fix transient product name while adding images.[https://github.com/woocommerce/woocommerce-ios/pull/5840]
+- [*] Simple Payments: Show tax breakup details in summary screen behind `taxLinesInSimplePayments` feature flag. [https://github.com/woocommerce/woocommerce-ios/pull/5943]
```
Since the feature is added behind a feature flag, it won't be accessible to users when released. I'd suggest only adding this to the release notes when the feature flag is enabled to always be `true`.

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/5943/commits/a9174fa9f70f155bc1d7cab85e9f6d66fc1722f6

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Simple Payments/Summary/SimplePaymentsSummary.swift:
```diff
@@ -339,25 +348,32 @@ private extension SimplePaymentsSummary {
             .previewDisplayName("Accessibility")
     }
+
+    static private func viewModel(noteContent: String? = nil) -> SimplePaymentsSummaryViewModel {
```
Nit: How about renaming this method to `createSampleViewModel` to clarify the intention?

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/5943/commits/dac7635902d0777e751a56a50208ab2e01d78143

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Simple Payments/Summary/SimplePaymentsSummaryViewModel.swift:
```diff
@@ -57,6 +80,22 @@ final class SimplePaymentsSummaryViewModel: ObservableObject {
+            return true // If `taxLinesInSimplePayments` feature flag is turned off
+        }
+
+        return taxLines.count > 0
```
Nit: checking for `taxLines.isNotEmpty` may be a bit more performant and straightforward I think.

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/5943/commits/68ae55ee6969edd9012adf4ba70ada69e899ae7a

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Simple Payments/Summary/SimplePaymentsSummaryViewModel.swift:
```diff
@@ -149,9 +196,18 @@ final class SimplePaymentsSummaryViewModel: ObservableObject {
+
+        // Generate `TaxLine`s to represent `taxes` inside `View`.
+        let taxLines = order.taxes.map({
+            TaxLine(id: $0.taxID,
+                    title: "\($0.label) (\($0.ratePercent)%)",
+                    value: currencyFormatter.formatAmount($0.totalTax) ?? $0.totalTax)
```
Nit: This logic can be simplified to be separate from `SimplePaymentsSummaryViewModel` in two ways:
- Use a convenience initializer for TaxLine that accepts `OrderTaxLine` as input
- Replace TaxLine with just an extension for OrderTaxLine which calculates the displayed title and value and conforms to Identifiable.

The separation of logic also makes it easier to test the computed variables too I think.

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/5943/commits/6dfb8324c504c879786ef80e24a3cfbded5e0b0b

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Simple Payments/SimplePaymentsSummaryViewModelTests.swift:
```diff
@@ -89,16 +103,176 @@ final class SimplePaymentsSummaryViewModelTests: XCTestCase {
+        XCTAssertTrue(viewModel.showChargeTaxesToggle)
+    }
+
+    func test_when_taxLines_is_not_empty_and_taxLinesInSimplePayments_feature_flag_is_off_then_showChargeTaxesToggle_is_true() {
```
Nit: this can be renamed to put the expected behavior before the conditions to make it clearer. This will make it easier to follow and consistent with the naming of other test cases:
```suggestion
    func test_showChargeTaxesToggle_is_true_when_taxLines_is_not_empty_and_taxLinesInSimplePayments_feature_flag_is_off() {
```

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/5943/commits/50f48dc47d31ae10c83b4727969eb2a89aba3fe3

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Simple Payments/SimplePaymentsSummaryViewModelTests.swift:
```diff
@@ -89,16 +103,176 @@ final class SimplePaymentsSummaryViewModelTests: XCTestCase {
+        XCTAssertTrue(viewModel.showChargeTaxesToggle)
+    }
+
+    func test_when_taxLines_is_empty_and_taxLinesInSimplePayments_feature_flag_is_on_then_showChargeTaxesToggle_is_false() {
```
Nit: Same comment as above - this can be renamed to be clearer:
```suggestion
    func test_showChargeTaxesToggle_is_false_when_taxLines_is_empty_and_taxLinesInSimplePayments_feature_flag_is_on() {
```

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/5943/commits/50f48dc47d31ae10c83b4727969eb2a89aba3fe3

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Simple Payments/SimplePaymentsSummaryViewModelTests.swift:
```diff
@@ -89,16 +103,176 @@ final class SimplePaymentsSummaryViewModelTests: XCTestCase {
+        XCTAssertFalse(viewModel.showChargeTaxesToggle)
+    }
+
+    func test_when_taxLines_is_not_empty_and_taxLinesInSimplePayments_feature_flag_is_on_then_showChargeTaxesToggle_is_true() {
```
Another nit 😄:
```suggestion
    func test_showChargeTaxesToggle_is_true_if_taxLines_is_not_empty_and_taxLinesInSimplePayments_feature_flag_is_on() {
```

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/5943/commits/50f48dc47d31ae10c83b4727969eb2a89aba3fe3

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Simple Payments/Summary/SimplePaymentsSummary.swift:
```diff
@@ -170,16 +170,25 @@ private struct PaymentsSection: View {
+                    .renderedIf(viewModel.showChargeTaxesToggle)
 
                 Group {
                     Text(SimplePaymentsSummary.Localization.taxesDisclaimer)
                         .footnoteStyle()
                         .padding(.horizontal)
                         .fixedSize(horizontal: false, vertical: true)
```
I found an edge case where you first have some tax set up in your store, turn on the Charge taxes toggle and then go back to remove the taxes. When creating a new simple payment, you'll still display this text:

<img src="https://user-images.githubusercontent.com/5533851/150723066-5f0aab40-8cbf-4dc8-bb5b-812aa7b90d06.png" width=320 />

Should this text be rendered only if `showChargeTaxesToggle` is true?

selanthiraiyan replied:
Great catch. 🎉 
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/5943/commits/41173b8cfcd032818cfa486bb75a895a9abe9101

itsmeichigo reviewed (changes requested):
This works well overall! I left a few comments and nit-picking below - let me know what you think.

Another minor issue that I found: when the tax line title is a bit long, the text seems to be center aligned, hence some extra space on the left of the text. Do you think making it align to the left will make it looks nicer?

<img src="https://user-images.githubusercontent.com/5533851/150723417-ff3d1da2-4891-47d4-9f19-c58d9a6f0ea3.png" width=320 />

Commit:
- Remove tax breakup details update from release notes. Should be added back only when feature flag is turned on for all.

Commit:
- Rename into `createSampleViewModel` for better clarity.

Commit:
- Replace `count > 0` check with  `isNotEmpty`.

Commit:
- Rename tests to move expected behaviour to the start of the method name.

Commit:
- Move `title` and `value` generation logic into `TaxLine`'s initializer.

Commit:
Fix the issue of showing taxes disclaimer even when `taxLines` are empty. (when `taxLinesInSimplePayments` flag is on)
- Show `taxesDisclaimer` text only when showing charge taxes toggle.

Commit:
- Introduce `showTaxesSection` to decide the visibility of taxes disclaimer text and tax line rows.
- Add tests to validate `showTaxesSection` behaviour.

Commit:
TitleAndValueRow `title`
- Use leading for `multilineTextAlignment` of `title` to make it look better where the text is long.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/TitleAndValueRow.swift:
```diff
@@ -18,6 +18,8 @@ struct TitleAndValueRow: View {
                 AdaptiveStack(horizontalAlignment: .leading) {
                     Text(title)
                         .style(bold: bold)
+                        .multilineTextAlignment(.leading)
```
> Another minor issue that I found: when the tax line title is a bit long, the text seems to be center aligned, hence some extra space on the left of the text. Do you think making it align to the left will make it looks nicer?

I thought of having a property `titleAlignment` (similar to`bold`) with `.center` as the default value  and passing in `.leading` from `PaymentsSection` View. But, `.center` isn't really the default value for `Text`. I think the text is aligned `center` due to the alignment logic in `AdaptiveStack`.

Also, I think that using `.leading` as the value for `.multilineTextAlignment` will suit all places `TitleAndValueRow` is used from.

 I am not proficient in SwiftUI so kindly let me know if you think that this isn't a good idea.

itsmeichigo replied:
> I think the text is aligned center due to the alignment logic in AdaptiveStack.

This seems to be correct! In the case where I found the issue, I was using a device with compact horizontal size class, so `AdaptiveStack` uses an HStack and we cannot really control the horizontal alignment in the stack directly. 

So, using `multilineTextAlignment` is the most sensible approach in my opinion. I'm wondering if we should use the same modifier on the value text as well, but with `.trailing` alignment to make sure the text looks good if it has more than one line.

Also, I'd suggest adding an extra view for the previews of the `TitleAndValueRow` view - to demonstrate the multi-line cases. WDYT?

selanthiraiyan commented:
@itsmeichigo Thank you for the detailed review and testing. 
I have made the necessary changes and the PR is ready for another look. 
Thank you!

selanthiraiyan commented:
Closing this to break it down into multiple PRs with the following changes,
- `TitleAndValueRow` - To make longer `title` text look nicer use `multilineTextAlignment` modifier with `leading` alignment. Check the multi line behaviour of `value` text and apply `multilineTextAlignment` with `trailing` alignment. Update SwiftUI preview to demo multi-line cases.
- Show `taxLines` from server behind `taxLinesInSimplePayments` feature flag. To avoid behaviour change, show Tax row with zero values if `taxLines` are empty.
